src/envs/doors.py

```diff
@@ -0,0 +1,933 @@
+                color = in_room_color
+            else:
+                color = default_room_color
+            room_geom.plot(ax, color=color)
+
+        # Draw robot.
+        robot_color = "blue"
+        robot_geom = self._object_to_geom(self._robot, state)
+        robot_geom.plot(ax, color=robot_color)
+
+        # Draw obstacles (including room walls).
+        obstacle_color = "black"
+        for obstacle in state.get_objects(self._obstacle_type):
+            obstacle_geom = self._object_to_geom(obstacle, state)
+            obstacle_geom.plot(ax, color=obstacle_color)
+
+        # Draw doors.
+        closed_door_color = "orangered"
+        open_door_color = "lightgreen"
+        doorway_color = "darkviolet"
+        for door in state.get_objects(self._door_type):
+            if self._DoorIsOpen_holds(state, [door]):
+                color = open_door_color
+            else:
+                color = closed_door_color
+            door_geom = self._object_to_geom(door, state)
+            door_geom.plot(ax, color=color)
+            if CFG.doors_draw_debug:
+                doorway_geom = self._door_to_doorway_geom(door, state)
+                doorway_geom.plot(ax, color=doorway_color, alpha=0.1)
+
+        # Visualize the motion plan.
+        if CFG.doors_draw_debug:
+            if action is not None and action.has_option():
+                option = action.get_option()
+                if "position_plan" in option.memory:
+                    xs, ys = zip(*option.memory["position_plan"])
+                    plt.scatter(xs, ys, s=3, alpha=0.5, color=robot_color)
+
+        pad = 2 * self.wall_depth
+        ax.set_xlim(x_lb - pad, x_ub + pad)
+        ax.set_ylim(y_lb - pad, y_ub + pad)
+
+        assert caption is None
+        plt.axis("off")
+        plt.tight_layout()
+        img = utils.fig2data(fig)
+        plt.close()
+        return [img]
+
+    def _get_tasks(self, num: int, rng: np.random.Generator) -> List[Task]:
+        tasks: List[Task] = []
+        for _ in range(num):
+            # Sample a room map.
+            room_map = self._sample_room_map(rng)
+            state = self._sample_initial_state_from_map(room_map, rng)
+            # Sample the goal.
+            rooms = state.get_objects(self._room_type)
+            candidate_goal_rooms = [
+                r for r in rooms
+                if not self._InRoom_holds(state, [self._robot, r])
+            ]
+            goal_room = candidate_goal_rooms[rng.choice(
+                len(candidate_goal_rooms))]
+            goal_atom = GroundAtom(self._InRoom, [self._robot, goal_room])
+            goal = {goal_atom}
+            assert not goal_atom.holds(state)
+            tasks.append(Task(state, goal))
+        return tasks
+
+    def _sample_initial_state_from_map(self, room_map: NDArray,
+                                       rng: np.random.Generator) -> State:
+        # Sample until a collision-free state is found.
+        while True:
```

Comment:
Seeing this makes me somewhat nervous that it'll never terminate. Looking at the logic, I see that that possibility is extremely unlikely, but maybe we want to include a timeout of some kind?